### ISSUE TITLE
TimescaleDB corrupt catalog tables

### ISSUE THREAD

**Issue by qiuyuhang** [timescale/timescaledb#2680]
**Relevant system information:**
 - OS: RedHat 7u2 3.10.0-327
 - PostgreSQL version (output of `postgres --version`): 12.2 (and all version support parallel worker)
 - TimescaleDB version (output of `\dx` in `psql`): master (and all version register hook: CacheRegisterSyscacheCallback, CacheRegisterRelcacheCallback)
 - Installation method: source

**Describe the bug**
ParallelWorkerMain will call InvalidateSystemCaches to RESET WORKER STATE，and hook of TimescaleDB will reproduce cache(by calling get_namespace_oid), which make parallel worker under wrong state, just like revert the postgres bug fix: https://github.com/postgres/postgres/commit/6de6d96d971c76c8d7bd3b31e6f182906137f913

One possible hook routine:
ParallelWorkerMain->InvalidateSystemCaches->inval_cache_callback(TimescaleDB)->extension_check->extension_current_state->proxy_table_exists->get_namespace_oid

It may also impact parallel index creating, parallel scan result (all routines use parallel worker). I haven't try it.

**To Reproduce**
1. preload timescaledb
2. create a lot of tables which contains a lot of columns（make it possible to trigger parallel worker on catalog tables, here is pg_attribute table, make it more than 100MB)
3. vacuum full pg_attribute
4. \d+ sometable, it's very likely to have pg_attribute corrupted.

**How this happens**
vacuum full pg_attribute will performDeletion(transient table)  after reindex_relation(pg_attribute) inside finish_heap_swap().
performDeletion(transient table) uses wrong index, so it delete wrong tuples.

**How to fix**
Use hash search like other hooks do, rather than get_namespace_oid.

**Expected behavior**
Noting happend.

**Actual behavior**
Corrupt catalog tables.

**Screenshots**
None

**Additional context**
None

**Comment by qiuyuhang** [comment]
I'm surprised that no one cares about this. This is a **FATAL** bug.

**Comment by erimatnor** [comment]
@qiuyuhang Thank you for the bug report. We are looking at this.

**Comment by qiuyuhang** [comment]
Here is the script.

```bash
export PGPORT=5432
export PGHOST=localhost
export PGDATABASE=bug2680
export PGUSER=postgres
export PGPASSWORD=postgres

psql -c "drop database $PGDATABASE" postgres
psql -c "create database $PGDATABASE" postgres

columnnum=1000
partnum=100
tablenum=10

columndef=""
for i in $(seq 0 $[$columnnum-1])
do
  columndef="$columndef,c$i int"
done

for i in $(seq 0 $[$tablenum-1])
do
  echo $i
  psql -c "drop table if exists userinfo$i" > /dev/null 2>&1
  psql -c "CREATE TABLE userinfo$i(id int$columndef) PARTITION BY HASH(id);" > /dev/null 2>&1
  for j in $(seq 0 $[$partnum-1])
  do
    psql -c "CREATE TABLE userinfo${i}_$j PARTITION of userinfo$i FOR VALUES WITH (MODULUS $partnum, REMAINDER $j);" > /dev/null 2>&1 &
  done
  wait < <(jobs -p)
done

psql -c "vacuum full analyze pg_attribute"
for i in $(seq 0 $[$tablenum-1])
do
  echo $i
  psql -c "\d userinfo$i" > /dev/null
done
```

**Comment by qiuyuhang** [comment]
"You are also indicating you know how to fix. Perhaps you can elaborate?"
Nope, I just give a possible way for I don't know what tsdb is doing inside these two hooks.
It seems no way to fix this by "Use hash search like other hooks do", namespace_oid is not a hash table.

After all, the basic idea is remove get_namespace_oid from CacheRegisterSyscacheCallback, CacheRegisterRelcacheCallback in parallel workers.

**Comment by qiuyuhang** [comment]
Can you reproduce this problem?

**Comment by NunoFilipeSantos** [comment]
Fixed on TimescaleDB 2.4.0.
